metadata-ingestion/src/datahub/ingestion/source/s3/source.py

```diff
@@ -771,18 +790,83 @@ def extract_table_data(
             )
         return table_data
 
+    def resolve_templated_folders(self, bucket_name: str, prefix: str) -> Iterable[str]:
+        folder_split = prefix.split("*", 1)
+        if len(folder_split) == 1:
+            yield prefix
+            return
+
+        folders = self.list_folders(bucket_name, folder_split[0])
+        for folder in folders:
+            yield from self.resolve_templated_folders(
+                bucket_name, f"{folder}{folder_split[1]}"
+            )
+
+    def list_folders(self, bucket_name: str, prefix: str) -> Iterable[str]:
+        assert self.source_config.aws_config
+        s3_client = self.source_config.aws_config.get_s3_client()
+        paginator = s3_client.get_paginator("list_objects_v2")
+        for page in paginator.paginate(
+            Bucket=bucket_name, Prefix=prefix, Delimiter="/"
+        ):
+            for o in page.get("CommonPrefixes", []):
+                folder: str = str(o.get("Prefix"))
+                if folder.endswith("/"):
+                    folder = folder[:-1]
+                yield f"{folder}"
+
     def s3_browser(self, path_spec: PathSpec) -> Iterable[Tuple[str, datetime, int]]:
         if self.source_config.aws_config is None:
             raise ValueError("aws_config not set. Cannot browse s3")
         s3 = self.source_config.aws_config.get_s3_resource()
         bucket_name = get_bucket_name(path_spec.include)
-        logger.debug(f"Scanning bucket : {bucket_name}")
+        logger.debug(f"Scanning bucket: {bucket_name}")
         bucket = s3.Bucket(bucket_name)
         prefix = self.get_prefix(get_bucket_relative_path(path_spec.include))
         logger.debug(f"Scanning objects with prefix:{prefix}")
-        for obj in bucket.objects.filter(Prefix=prefix).page_size(1000):
-            s3_path = f"s3://{obj.bucket_name}/{obj.key}"
-            yield s3_path, obj.last_modified, obj.size,
+        matches = re.finditer(r"{\s*\w+\s*}", path_spec.include, re.MULTILINE)
+        matches_list = list(matches)
+        if len(matches_list) > 0 and path_spec.sample_files:
+            # Replace the patch_spec include's templates with star because later we want to resolve all the stars
+            # to actual directories.
+            # For example:
+            # "s3://my-test-bucket/*/{dept}/*/{table}/*/*.*" -> "s3://my-test-bucket/*/*/*/{table}/*/*.*"
+            # We only keep the last template as a marker to know the point util we need to resolve path.
+            # After the marker we can safely get sample files for sampling because it is not used in the
+            # table name, so we don't need all the files.
+            # This speed up processing but we won't be able to get a precise modification date/size/number of files.
+            max_start = -1
+            include = path_spec.include
+            max_match = ""
+            for match in matches_list:
+                pos = include.find(match.group())
+                if pos > max_start:
```

Comment:
looks like `max_start` is never getting updated. Is it always the constant `-1`?